staging/src/k8s.io/apimachinery/pkg/runtime/scheme.go

```diff
@@ -620,3 +628,69 @@ func setTargetKind(obj Object, kind schema.GroupVersionKind) {
 	}
 	obj.GetObjectKind().SetGroupVersionKind(kind)
 }
+
+// this only fails on programmer error
+func (s *Scheme) SetVersionPriorityOrDie(versions ...schema.GroupVersion) {
+	if err := s.SetVersionPriority(versions...); err != nil {
+		panic(err)
+	}
+}
+
+func (s *Scheme) SetVersionPriority(versions ...schema.GroupVersion) error {
+	groups := sets.String{}
+	order := []string{}
+	for _, version := range versions {
+		groups.Insert(version.Group)
+		order = append(order, version.Version)
+	}
+	if len(groups) != 1 {
+		return fmt.Errorf("must register versions for exactly one group: %v", strings.Join(groups.List(), ", "))
+	}
+
+	s.versionPriority[groups.List()[0]] = order
+	return nil
+}
+
+func (s *Scheme) PrioritizedVersionsForGroup(group string) []schema.GroupVersion {
+	ret := []schema.GroupVersion{}
+	for _, version := range s.versionPriority[group] {
+		ret = append(ret, schema.GroupVersion{Group: group, Version: version})
+	}
+	return ret
+}
+
+func (s *Scheme) PrioritizedVersionsAllGroups() []schema.GroupVersion {
+	ret := []schema.GroupVersion{}
+	for group, versions := range s.versionPriority {
+		for _, version := range versions {
+			ret = append(ret, schema.GroupVersion{Group: group, Version: version})
+		}
+	}
+	return ret
+}
+
+func (s *Scheme) PreferredVersionsAllGroups() []schema.GroupVersion {
+	ret := []schema.GroupVersion{}
+	for group, versions := range s.versionPriority {
+		for _, version := range versions {
+			ret = append(ret, schema.GroupVersion{Group: group, Version: version})
+			break
+		}
+	}
+	return ret
+}
+
+func (s *Scheme) IsGroupRegistered(group string) bool {
```

Comment:
the following seems very strange to me:

```go
gv := schema.GroupVersion{Group:"foo",Version:"v1"}
scheme.AddKnownTypes(gv, ...)
scheme.KnownTypes(gv) // returns results
scheme.IsGroupRegistered(gv.Group) // returns false
```

if you register types for a few versions of a group and do not set an explicit priority order, then ask for the priority order, I'd expect either to return the first seen order of the versions, or error. I wouldn't expect the returned list to not contain the group/versions the scheme knew about

likewise, if I register KnownTypes for 4 versions of a group, then set an explicit priority of 2 of those versions, I'd either expect the other two versions to go at the end in indeterminate order, or to error. I wouldn't expect the other 2 versions to disappear from the ordered list of all versions.